File: src/FSharpSolutions/MutableState/DiscountCalculator.fs
namespace FunctionalKata.MutableState

type DiscountCalculator(
    minimumFlatDiscountAmount: decimal,
    minimumBulkDiscountQuantity: int,
    flatDiscountRate: decimal,
    bulkDiscountRate: decimal) =

    new() = DiscountCalculator(1000m, 10, 0.1m, 0.05m)

    member this.CalculateDiscount(cart: ShoppingCart) : decimal =
        let flatDiscount =
            if cart.TotalAmount >= minimumFlatDiscountAmount then
                cart.TotalAmount * flatDiscountRate
            else
                0m
        
        let bulkDiscount =
            cart.Items
            |> Seq.sumBy (fun item ->
                if item.Quantity >= minimumBulkDiscountQuantity then
                    item.Price * decimal item.Quantity * bulkDiscountRate
                else
                    0m)
        
        flatDiscount + bulkDiscount

File: src/FunctionalKataSolutions/MutableState/DiscountCalculator.cs
namespace FunctionalKata.MutableState;

public class DiscountCalculator(
    decimal minimumFlatDiscountAmount = 1000,
    int minimumBulkDiscountQuantity = 10,
    decimal flatDiscountRate = 0.1m,
    decimal bulkDiscountRate = 0.05m)
{

    public decimal CalculateDiscount(ShoppingCart cart) =>
        CalculateFlatRateDiscount(cart) + CalculateBulkDiscount(cart);

    private decimal CalculateFlatRateDiscount(ShoppingCart cart) =>
        cart.TotalAmount >= minimumFlatDiscountAmount
            ? cart.TotalAmount * flatDiscountRate
            : 0.0m;

    private decimal CalculateBulkDiscount(ShoppingCart cart) =>
        cart.Items
            .Where(item => item.Quantity >= minimumBulkDiscountQuantity)
            .Sum(item => item.Price * item.Quantity * bulkDiscountRate);
    
}